Simulator/simulator.py

```diff
@@ -95,324 +85,687 @@ def __init__(self, etype: etype, time: datetime, job_id: int):
+            waiting_time_threshold_hour: int,
+    ) -> None:
+        max_wait_time_sec = waiting_time_threshold_hour * SECONDS_IN_HOUR
+        runtime_threshold_sec = runtime_threshold_min * 60
+        self.input_trace['speculative_execution'] = False
+
+        next_event = self._get_next_event()
+        model = self.train_waiting_time_model(max_wait_time_sec)
+        total_on_demand_cost = 0
+
+        run_immediately = 0
+        run_spec_exec_correct = 0
+        run_spec_exec_incorrect = 0
+        waited_in_queue = 0
+        waited_max_time = 0
+
+        while next_event:
+            if next_event.etype == etype.schedule:
+                # Schedule new job
+
+                cur_job = self.input_trace.loc[self.input_trace['job_id'] == next_event.job_id].iloc[0]
+                cur_job_id, cur_time, cpu_alloc, runtime = get_job_attributes(cur_job)
+
+                if self.queue_is_empty() and self._can_schedule(cpu_alloc):
+                    # Run the job on fixed resources - no queue & sufficient resources available
+                    self._schedule(cur_time, cpu_alloc, runtime, cur_job_id)
+                    self.set_waiting_time(cur_job, 0)
+                    run_immediately += 1
+                elif is_long_job(cur_job, runtime_threshold_min) and self.is_short_wait(model, cur_job):
+                    # Jobs should wait for fixed resources if they are 1) long or 2) have short waits
+                    self.wait_queue.append(cur_job.job_id)
+                elif cur_job.speculative_execution:
+                    self.wait_queue.append(cur_job.job_id)
+                else:
+                    # Run the job on on-demand -- short job or long wait
+                    spec_exec_runtime = min(runtime_threshold_sec, runtime)
+                    total_on_demand_cost += self.calc_on_demand_cost_for_job(cur_job, spec_exec_runtime)
+                    if runtime <= runtime_threshold_sec:
+                        self.set_waiting_time(cur_job, 0)
+                        run_spec_exec_correct += 1
+                    else:
+                        self.resubmit_long_jobs_after_spec_execution(cur_job, runtime_threshold_sec)
+                        run_spec_exec_incorrect += 1
+
+            else:
+                # A job is finished. So, check the wait queue
+                while len(self.wait_queue) > 0:
+                    waiting_job = self.input_trace.loc[self.input_trace['job_id'] == self.wait_queue[0]].iloc[0]
+                    waiting_job_id, waiting_job_time, waiting_job_cpu_alloc, waiting_job_runtime = \
+                        get_job_attributes(waiting_job)
+                    time_waited_sec = next_event.time - waiting_job_time + \
+                                      (waiting_job.speculative_execution * runtime_threshold_sec)
+
+                    # Check if job waited more than the max waiting time
+                    if time_waited_sec > (max_wait_time_sec):
```

Comment:
for my own understanding here. If time_waited_sec > max_wait_time_sec, it means that job would've been run on demand. It will be run sometime before the current job finishes, but we just do the bookeeping here, right?

Reply:
hmm let me know if this answers your question. this part of the code isn't checking whether the wait is predicted to be short under SWW - it's instead making sure that no job waits more than the waiting time threshold "b". so, if the first job in the wait queue has been waiting for more than "b" time (e.g. 6 hours), the simulator removes it from the queue and runs it on-demand instead. this is just to prevent situations where a job is waiting for far too long and clogs up the queue

Reply:
Yeap I think your explanation is a better way of wording my assumption!